### ISSUE TITLE
[FEATURE] Replace the polylabel dependency with an own implementation

### ISSUE THREAD

**Issue by josxha** [fleaflet/flutter_map#1746]
### What do you want implemented?

The current integration of polylabel is not very well performance optimized. By implementing our own port of polylabel, we can:
-  have better performance
- decrease our dependance on 3rd party packages

### Can you provide any other information?

Current usage of the `polylabel` package:

```dart
LatLng _computePolylabel(List<LatLng> points) {
  final labelPosition = polylabel(
    [
      List<math.Point>.generate(points.length,
          (i) => math.Point(points[i].longitude, points[i].latitude)),
    ],
    // "precision" is a bit of a misnomer. It's a threshold for when to stop
    // dividing-and-conquering the polygon in the hopes of finding a better
    // point with more distance to the polygon's outline. It's given in
    // point-units, i.e. degrees here. A bigger number means less precision,
    // i.e. cheaper at the expense off less optimal label placement.
    precision: 0.000001,
  );
  return LatLng(
    labelPosition.point.y.toDouble(),
    labelPosition.point.x.toDouble(),
  );
}
```

### Severity

Minimum: Not required for my use

**Comment by JosefWN** [comment]
Maybe better to be inspired rather than straight-up porting, perhaps a bit touchy since it's a commercial company, and porting counts as redistribution from a legal point of view. In their licensing terms:
```
...
* Redistributions of source code must retain the above copyright notice,
  this list of conditions and the following disclaimer.
* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.
...
  ```

Only dependencies are managed automatically by flutter (bundling the licenses with the binary), so it could be a bit cumbersome 😅

**Comment by JaffaKetchup** [comment]
(It is possible to use `LicenseBinding` to add licenses to the Flutter app: https://api.flutter.dev/flutter/foundation/LicenseRegistry/addLicense.html)

**Comment by JosefWN** [comment]
Ah, thanks, didn't know! If we could automate that it would certainly simplify things for the users of `flutter_map`, I'm still doubtful that porting is allowed within the scope of the TOS though?

**Comment by josxha** [comment]
> I'm still doubtful that porting is allowed within the scope of the TOS though

Good point, in that case we should see that as an encouragement to remove the dependency to the polylabel package anyways.

**Comment by navispatial** [comment]
Maybe a late comment on TOS problems mentioned above but MapBox has published their `polylabel` implementation as JavaScript code in a separate repository: https://github.com/mapbox/polylabel

The license https://github.com/mapbox/polylabel/blob/master/LICENSE should allow any kind of ports (however this is not a legally verified comment...) as long as license text is provided in all copies.

This is the JavaScript file that covers the whole `polylabel` algorithm: https://github.com/mapbox/polylabel/blob/master/polylabel.js

It has a dependency to a very simple TinyQueue implementation (with a very permissive license too):
https://github.com/mourner/tinyqueue/blob/main/index.js

Instead of that it's possible to use `PriorityQueue` class from the Dart `package:collection` package (as `flutter_map` seems to already have a dependency on it).  

There's also an another port of `polylabel` to Dart in the `geobase` package, but maybe that's not an answer either to this issue (as transformation between polyline data structures would be needed also when using it):
* https://github.com/navibyte/geospatial/blob/main/dart/geobase/lib/src/geometric/cartesian/areal/polylabel.dart
* https://pub.dev/packages/geobase

**Comment by josxha** [comment]
Hello @navispatial, thanks for your message!
I'd be interested in what your ongoing plans with `geobase` are? I've seen that geobase is not limited to any projection but supports simple geometries as well, just like flutter_map intends to be. With recent versions, you've added a lot of more complex algorithms like polylabel, point-in-polygon, vincentry and so on. It's great work and you have great documentation as well!
I'd like to ask if you have plans to some day have most or all functionality geospartial analysis tools like [turfjs](https://turfjs.org/) provide? Another more general question, is geobase [RFC 7946](https://datatracker.ietf.org/doc/html/rfc7946) compliant, and will it be in the future?

**Comment by navispatial** [comment]
Hello @josxha! Thanks for comments and questions on the `geobase` package . 

I just updated an issue [Roadmap 2025: The current state and candidate issues for the geobase package #254](https://github.com/navibyte/geospatial/issues/254) on the [geospatial](https://github.com/navibyte/geospatial) repository. Maybe that could give hints what are plans on `geobase`. Next release `geobase 1.5.0` is not scheduled yet but maybe February or March 2025 might be a good guess.

About your questions:

RFC 7946 compliance:
* [simple geometry](https://geospatial.navibyte.dev/v1/geobase/geometry/) and [geospatial feature](https://geospatial.navibyte.dev/v1/geobase/features/) classes should be able to represent all data structures decoded from a RFC 7946 compliant GeoJSON data
* also [encoding / decoding](https://geospatial.navibyte.dev/v1/geobase/formats/#geojson) support for GeoJSON is based on the RFC 7946
* there are internal tests on this
* however no official compliance tests on RFC 7946 are applied

Geospatial analysis tools like those provided by `turfjs`
* few new 2D cartesian algorithms and UTM / MGRS based functions on ellipsoidal geodesy is planned on the roadmap of `geobase`
* however `turfjs` is quite extensive library based on WGS84 geographic coordinates
* currently (at least for 2025) there are no plans on `geobase` to support such extensive set of geospatial analysis functions (even if a small set of similar functionality already supported by `geobase`)
